### PR TITLE
ucm2: Add initial support for AMD Vangogh (acp5x) on Steam Deck

### PR DESCRIPTION
**Pull request by cristicc**
This has been tested on a Valve Steam Deck EV2 unit, using kernel v6.1-rc1.

### CODE REVIEW

File: ucm2/AMD/acp5x/acp5x.conf [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Syntax 3
```
Comment (perexg) [inline ucm2/AMD/acp5x/acp5x.conf:1]: You should use DMI or any other information to use this acp5x configuration only for this specific hardware.

Comment (cristicc) [reply ucm2/AMD/acp5x/acp5x.conf:1]: Thanks for reviewing!

I have just pushed an updated version, restricting the usage to the Steam Deck hardware.

**Review by perexg — CHANGES_REQUESTED** [review]

File: ucm2/AMD/acp5x/HiFi.conf [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		cset "name='Right AMP Enable Switch' off"
+	]
+
+	Value {
+		PlaybackPriority 100
+		PlaybackPCM "hw:acp5x,1"
+	}
+}
+
+SectionDevice."Internal Mic".0 {
```
[the remark this answers is not in the document]
Comment (cristicc) [reply ucm2/AMD/acp5x/HiFi.conf]: Done.

File: ucm2/AMD/acp5x/HiFi.conf [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	]
+
+	DisableSequence [
+		cset "name='Left AMP Enable Switch' off"
+		cset "name='Right AMP Enable Switch' off"
+	]
+
+	Value {
+		PlaybackPriority 100
+		PlaybackPCM "hw:acp5x,1"
```
[the remark this answers is not in the document]
Comment (cristicc) [reply ucm2/AMD/acp5x/HiFi.conf]: Done.

File: ucm2/AMD/acp5x/HiFi.conf [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+	EnableSequence [
+		cset "name='Left DSP1 Preload Switch' 1"
+		cset "name='Right DSP1 Preload Switch' 1"
+		cset "name='Left PCM Source' DSP"
+		cset "name='Right PCM Source' DSP"
+	]
+
+	DisableSequence [
+		cset "name='Left AMP Enable Switch' off"
```
[the remark this answers is not in the document]
Comment (cristicc) [reply ucm2/AMD/acp5x/HiFi.conf]: Fixed.

File: ucm2/AMD/acp5x/HiFi.conf [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	Comment "Internal Microphone"
+
+	ConflictingDevice [
+		"Headset"
+	]
+
+	EnableSequence [
+		cset "name='Int Mic Switch' on"
+		cset "name='DMIC Enable Switch' on"
+		cset "name='Mic Volume' 252"
```
[the remark this answers is not in the document]
Comment (cristicc) [reply ucm2/AMD/acp5x/HiFi.conf]: User configurable settings moved to BootSequence.

File: ucm2/AMD/acp5x/HiFi.conf [hunk 6]
```diff
@@ -0,0 +1,122 @@
+SectionVerb {
+	EnableSequence [
+		cset "name='Left DSP1 Preload Switch' 0"
```
[the remark this answers is not in the document]
Comment (cristicc) [reply ucm2/AMD/acp5x/HiFi.conf]: Many thanks for the detailed review and the hints to further improve this! Hopefully I got it in a better shape this time..

**Comment by DocMAX** [comment]
Works perfectly here. What are we still waiting for?...

**Comment by bell07** [comment]
The "off" Profile does not work as expected for me. The output devices are still available if I try to disable the port.

Steps to reproduce: Connect steam deck to TV. Play audio to builtin (acp5x) speaker by pulseaudio / pipewire. 
Try to disable the the buitin speaker by
`pactl set-card-profile alsa_card.pci-0000_04_00.5-platform-acp5x_mach.0 off`

expected behaviour is the both acp5x output devices are not available and audio switch to HDMI output.

Currently I use the

**Comment by cristicc** [comment]
Thanks @DocMAX and @bell07 for testing this!

I've just pushed an update to properly handle the "off" profile. Please let me know if you still encounter issues.

**Comment by bell07** [comment]
Thank you! Does perfectly work for me now

**Comment by hexaheximal** [comment]
I tried copying the files in this pull request to my Steam Deck (which is currently running Debian bookworm/testing), and it did not show the HiFi option in pavucontrol.

I additionally tried running `alsaucm reload`, and it showed multiple errors:

```
ALSA lib parser.c:244:(error_node) UCM is not supported for this HDA model (HD-Audio Generic at 0x803c0000 irq 72)
ALSA lib main.c:1541:(snd_use_case_mgr_open) error: failed to import hw:0 use case configuration -6
alsaucm: error failed to open sound card hw:0: No such device or address
```

I have rebooted it multiple times, but it still doesn't show the HiFi option and only shows the HDMI options.

What configuration step am I missing?

**Comment by cristicc** [comment]
@hexaheximal Although I haven't encountered that kind of issues so far, I would suggest to follow the installation procedure documented in [README.md](https://github.com/alsa-project/alsa-ucm-conf/blob/master/README.md), just to exclude a possible configuration problem. Make sure to do a backup first, since this might break your setup if the installed ALSA binaries are too old and cannot properly handle the configuration update.

**Comment by hexaheximal** [comment]
> @hexaheximal Although I haven't encountered that kind of issues so far, I would suggest to follow the installation procedure documented in [README.md](https://github.com/alsa-project/alsa-ucm-conf/blob/master/README.md), just to exclude a possible configuration problem. Make sure to do a backup first, since this might break your setup if the installed ALSA binaries are too old and cannot properly handle the configuration update.

I'm using the version of alsa-ucm-conf (and the kernel) included in debian bookworm.
I don't see why it wouldn't work though.
I've heard it only works on kernel 6.1+ before, but I can't find anything that changed in 6.1 for the acp5x driver.

**Comment by cristicc** [comment]
> I'm using the version of alsa-ucm-conf (and the kernel) included in debian bookworm.

I only tested on an Arch Linux based distro after updating alsa-ucm-conf as indicated above.

> I've heard it only works on kernel 6.1+ before, but I can't find anything that changed in 6.1 for the acp5x driver.

That's because not all the changes are on the driver itself, e.g. you need at least the following patch from 6.1 to get proper sound support:
```
9d08f700ab78 spi: amd: Setup all xfers before opcode execution
```

**Comment by waffshappen** [comment]
I am using this on a Steam Deck with Fedora 37 on the latest 6.1.2 kernel from koji.

While the patchset works generally and makes the speakers work they are very, very hushed and in any slightly louder environment almost completely inaudible at max volume (125% boost, gnome+pipewire). Booting back to SteamOS they're at the expected max loudness again. Is this expected for now?

**Comment by cristicc** [comment]
@waffshappen Yes, this is expected. 

SteamOS achieves max loudness by using a modified/tuned firmware for the cs35l41 audio codec. You may try to locate `cs35l41-dsp1-spk-prot.wmfw` and `cs35l41-dsp1-spk-prot.bin` on your Fedora distro (most probably they reside in `/lib/firmware/cirrus`) and replace them with `/lib/firmware/{cs35l41-dsp1-spk-prot.wmfw,cs35l41-dsp1-spk-prot.bin}` provided by SteamOS.

WARNING: Before doing any audio test with the replaced firmware, make sure to lower the volume first, otherwise you may risk damaging the speakers! On the SteamOS side, this is mitigated via the UCM configuration by restricting user access to some mixer controls.

**Comment by waffshappen** [comment]
> @waffshappen It is expected to get a slightly lower loudness, but not quite to the levels you describe. Probably some of the mixer controls are not properly adjusted.
> Also note SteamOS achieves max loudness by using a modified/tuned firmware for the cs35l41 audio codec. You may try to locate `cs35l41-dsp1-spk-prot.wmfw` and `cs35l41-dsp1-spk-prot.bin` on your Fedora distro (most probably they reside in `/lib/firmware/cirrus`) and replace them with `/lib/firmware/{cs35l41-dsp1-spk-prot.wmfw,cs35l41-dsp1-spk-prot.bin}` provided by SteamOS.
> WARNING: Before doing any audio test with the replaced firmware, make sure to lower the volume first, otherwise you may risk damaging the speakers! On the SteamOS side, this is mitigated via the UCM configuration by restricting user access to some mixer controls.

I tested this with both the original steam deck and fedora firmware - both were initially equally silent until i found (alsamixer, disabled pw alsa redirection) the "Left Analog PCM" (and right variant) and set both to dB gain: 0,20

This made loudness equal to the original steam deck loudness on both firmware versions. Maybe something to adjust in the ucm config, if possible?

**Comment by cristicc** [comment]
> This made loudness equal to the original steam deck loudness on both firmware versions. Maybe something to adjust in the ucm config, if possible?

The adjustment is already handled in acp5x.conf:
```
BootSequence [
	...
	cset "name='Left Analog PCM Volume' 17"
	cset "name='Right Analog PCM Volume' 17"
	...
]
```
Both controls should have been automatically set to 0.17dB if the sound card state was not previously saved, i.e. file `/var/lib/alsa/asound.state` is missing. If you don't have this file, probably the mixer controls have been adjusted via pulse/pipewire?! 

After rebooting the system, what does alsamixer show? I would expect either 0.20 (what you manually set) or 0.17 (what UCM provides as defaults).

**Comment by waffshappen** [comment]
> > This made loudness equal to the original steam deck loudness on both firmware versions. Maybe something to adjust in the ucm config, if possible?
> 
> The adjustment is already handled in acp5x.conf:
> 
> ```
> BootSequence [
> 	...
> 	cset "name='Left Analog PCM Volume' 17"
> 	cset "name='Right Analog PCM Volume' 17"
> 	...
> ]
> ```
> 
> Both controls should have been automatically set to 0.17dB if the sound card state was not previously saved, i.e. file `/var/lib/alsa/asound.state` is missing. If you don't have this file, probably the mixer controls have been adjusted via pulse/pipewire?!
> 
> After rebooting the system, what does alsamixer show? I would expect either 0.20 (what you manually set) or 0.17 (what UCM provides as defaults).

It is still on 0.20 as was set via alsamixer.

Could this be a sideeffect of having to install the distro first and then adding the configs and kernel 6.1 after first boot?

**Comment by cristicc** [comment]
> Could this be a sideeffect of having to install the distro first and then adding the configs and kernel 6.1 after first boot?

On Arch Linux (and most likely on Fedora as well), the [alsa-utils](https://archlinux.org/packages/?name=alsa-utils) package provides the `alsa-restore.service` responsible for saving the sound card state to `/var/lib/alsa/asound.state` on system shutdown and restoring it after booting:
```
$ systemctl status alsa-restore
● alsa-restore.service - Save/Restore Sound Card State
     Loaded: loaded (/usr/lib/systemd/system/alsa-restore.service; static)
[...]
Jan 08 22:37:11 systemd[1]: Starting Save/Restore Sound Card State...
Jan 08 22:37:11 systemd[1]: Finished Save/Restore Sound Card State.
```
The most plausible scenario is that the sound card state had been already persisted when you updated the UCM configuration, hence the `BootSequence` in `acp5x.conf` did not execute.

**Comment by perexg** [comment]
You may remove the old state using:

```
alsactl stop alsa-state
alsactl stop alsa-restore
rm /var/lib/alsa/asound.state
reboot
```

And yes, `BootSequence` is executed only when the previous state does not exist.

**Comment by cristicc** [comment]
@perexg Is there anything else you'd like me to address/improve in order to get this merged?

**Pull request closed without merging**